### ISSUE TITLE
webdav

### ISSUE THREAD

**Issue by fdefilippo** [drakkan/sftpgo#207]
hello, I use sftpgo as a bridge to an s3 endpoint that when a file is uploaded moves it elsewhere. Using webdav and uploading a file with windows the explorer say: 

![explorer_nfqgfjM0YX](https://user-images.githubusercontent.com/16714986/97302227-10539080-1859-11eb-8060-a1afe8e92eb6.png)

this happens because the file is no longer avilable and I guess it is done a stat that fails. Using Curl,for obvious reasons, is ok :

```
> PUT /fake0/out/test.bin HTTP/1.1
> Authorization: Basic XYZ
> User-Agent: curl/7.29.0
> Host: FQDN:2080
> Accept: */*
> Content-Length: 441
> Expect: 100-continue
>
< HTTP/1.1 100 Continue
* We are completely uploaded and fine
< HTTP/1.1 201 Created
< Etag: "1641d9285d8c3d961b9"
< Date: Tue, 27 Oct 2020 12:28:29 GMT
< Content-Length: 7
< Content-Type: text/plain; charset=utf-8
<
* Connection #0 to host FQDN left intact

```

could it be a cache problem (assuming you use it) of the driver that uses sftpgo?

thnx a lot.

**Comment by drakkan** [comment]
Hi,

there is no cache, but the webdav implementation does a lot of stat calls (some are internal, other are explicitly requested from the client). 

Similar issues happen if you remove a directory tree via webdav using a cloud storage as backed, since the directory are virtual they are automatically removed and the client complains that they are missing (nautilus try to recursively remove each file and folder, not tested with IE). 

Basically, in the current status, webdav does not work well with cloud backends.

In your case you should try to understand if the stat is originated client side or if it is server side, thank you

**Comment by fdefilippo** [comment]
Hi, it's likely a client-side problem or at least it seems, below the dialogue: 
```
{"level":"info","time":"2020-10-27T14:58:05.210","sender":"webdavd","elapsed_ms":12,"method":"PROPFIND","proto":"HTTP/2.0","remote_addr":"127.0.0.1:44004","request_id":"buc2eradsk96afmlit7g","uri":"https://localhost:2080/test/out","user_agent":"Microsoft-WebDAV-MiniRedir/10.0.19041"}
{"level":"info","time":"2020-10-27T14:58:05.431","sender":"Upload","elapsed_ms":134,"size_bytes":0,"username":"test","file_path":"/out/datafile.txt","connection_id":"DAV_buc2eradsk96afmlit80","protocol":"DAV"}
{"level":"info","time":"2020-10-27T14:58:05.431","sender":"webdavd","elapsed_ms":142,"method":"PUT","proto":"HTTP/2.0","remote_addr":"127.0.0.1:44004","request_id":"buc2eradsk96afmlit80","uri":"https://localhost:2080/test/out/datafile.txt","user_agent":"Microsoft-WebDAV-MiniRedir/10.0.19041"}
{"level":"info","time":"2020-10-27T14:58:05.482","sender":"webdavd","elapsed_ms":4,"method":"LOCK","proto":"HTTP/2.0","remote_addr":"127.0.0.1:44004","request_id":"buc2eradsk96afmlit8g","uri":"https://localhost:2080/test/out/datafile.txt","user_agent":"Microsoft-WebDAV-MiniRedir/10.0.19041"}
{"level":"info","time":"2020-10-27T14:58:05.529","sender":"webdavd","elapsed_ms":0,"method":"PROPPATCH","proto":"HTTP/2.0","remote_addr":"127.0.0.1:44004","request_id":"buc2eradsk96afmlit90","uri":"https://localhost:2080/test/out/datafile.txt","user_agent":"Microsoft-WebDAV-MiniRedir/10.0.19041","error":"webdav: locked"}
{"level":"info","time":"2020-10-27T14:58:05.614","sender":"webdavd","elapsed_ms":5,"method":"HEAD","proto":"HTTP/2.0","remote_addr":"127.0.0.1:44004","request_id":"buc2eradsk96afmlit9g","uri":"https://localhost:2080/test/out/datafile.txt","user_agent":"Microsoft-WebDAV-MiniRedir/10.0.19041"}
{"level":"info","time":"2020-10-27T14:58:05.658","sender":"webdavd","elapsed_ms":0,"method":"PUT","proto":"HTTP/2.0","remote_addr":"127.0.0.1:44004","request_id":"buc2eradsk96afmlita0","uri":"https://localhost:2080/test/out/datafile.txt","user_agent":"Microsoft-WebDAV-MiniRedir/10.0.19041","error":"webdav: locked"}
{"level":"info","time":"2020-10-27T14:58:05.704","sender":"webdavd","elapsed_ms":0,"method":"DELETE","proto":"HTTP/2.0","remote_addr":"127.0.0.1:44004","request_id":"buc2eradsk96afmlitag","uri":"https://localhost:2080/test/out/datafile.txt","user_agent":"Microsoft-WebDAV-MiniRedir/10.0.19041","error":"webdav: locked"}
{"level":"info","time":"2020-10-27T14:58:06.085","sender":"Upload","elapsed_ms":318,"size_bytes":0,"username":"test","file_path":"/out/datafile.txt","connection_id":"DAV_buc2eradsk96afmlitb0","protocol":"DAV"}
{"level":"info","time":"2020-10-27T14:58:06.085","sender":"webdavd","elapsed_ms":327,"method":"PUT","proto":"HTTP/2.0","remote_addr":"127.0.0.1:44004","request_id":"buc2eradsk96afmlitb0","uri":"https://localhost:2080/test/out/datafile.txt","user_agent":"Microsoft-WebDAV-MiniRedir/10.0.19041"}
```

**Comment by fdefilippo** [comment]
hi, another problem, querying, PROPFIND, a bucket with 670 objects takes several minutes to return the output, doing the same with the s3 client the output is near realtime, the difference is in the number of requests: in the case of the s3 client there are about ten , in the case of PROPFIND they are countless.

Number of s3 objects:
<pre>
$ time s3mc ls --recursive local| wc -l
681

real 0m0.619s
user 0m0.171s
sys 0m0.074s

and 9 request to the s3 server.
</pre>

curl:
<pre>
$ time curl -ksS -X PROPFIND -u test:xyz -o /dev/null https://localhost:2080/test

real    7m14.335s
user    0m0.050s
sys     0m0.019s

and 3139 request to the s3 server:
$ wc -l propfind.log
3139 propfind.log

</pre>

so the client side of sftpgo do a lot of call. one last example, just to list a prefix with two objects it makes 13 requests: 

curl -ksS -u text:xyz https://localhost:2080/test/reject/

<pre>
s3.ListObjectsV2 s3.lab.lan:9000/test?delimiter=%2F&list-type=2&prefix= 
s3.ListObjectsV2 s3.lab.lan:9000/test?delimiter=%2F&list-type=2&prefix=
s3.ListObjectsV2 s3.lab.lan:9000/test?delimiter=%2F&list-type=2&prefix= 
s3.ListObjectsV2 s3.lab.lan:9000/test?delimiter=%2F&list-type=2&prefix=
s3.ListObjectsV2 s3.lab.lan:9000/test?delimiter=%2F&list-type=2&prefix=
s3.ListObjectsV2 s3.lab.lan:9000/test?delimiter=%2F&list-type=2&prefix=reject%2F
s3.ListObjectsV2 s3.lab.lan:9000/test?delimiter=%2F&list-type=2&prefix=reject%2F
s3.ListObjectsV2 s3.lab.lan:9000/test?delimiter=%2F&list-type=2&prefix=reject%2F
s3.ListObjectsV2 s3.lab.lan:9000/test?delimiter=%2F&list-type=2&prefix=reject%2F
s3.ListObjectsV2 s3.lab.lan:9000/test?delimiter=%2F&list-type=2&prefix=reject%2F
s3.ListObjectsV2 s3.lab.lan:9000/test?delimiter=%2F&list-type=2&prefix=reject%2F 
s3.ListObjectsV2 s3.lab.lan:9000/test?delimiter=%2F&list-type=2&prefix=reject%2F  
s3.HeadObject s3.lab.lan:9000/test/reject/ucs.license
</pre>

thanks.

**Comment by drakkan** [comment]
Hi,

if you use SFTP/FTP is the number of requests to S3 similar to the one you get using `s3mc ls`? WebDAV need the mimetype, this means an additional HEAD request for each object. SFTPGo tries to avoid this additional request guessing the [mimetye](https://github.com/drakkan/sftpgo/blob/master/webdavd/file.go#L60). Is you file extension listed in `/etc/mime.types`? If this fail an head request will be performed, if the HEAD request fails too and SFTPGo returns `webdav.ErrNotImplemented` the webdav library will read some bytes from the file to guess the mime type. Can you please try to understand what is happening in your case? thank you

**Comment by fdefilippo** [comment]
Hi, if I use sftp/ftp the number or requests are the same as s3mc ls. Ok the mime type are unknown so sftpgo try to recognize it. But I did a second test using a webdav server (https://github.com/hacdias/webdav) and s3fs by mounting the bucket on the local file system, the result is much faster (few seconds vs several minutes) it make only s3.HeadObject calls to s3. What can I do to help you understand any problem?

**Comment by drakkan** [comment]
> Hi, if I use sftp/ftp the number or requests are the same as s3mc ls. Ok the mime type are unknown so sftpgo try to recognize it. But I did a second test using a webdav server (https://github.com/hacdias/webdav) and s3fs by mounting the bucket on the local file system, the result is much faster (few seconds vs several minutes) it make only s3.HeadObject calls to s3. What can I do to help you understand any problem?

This could be a different test, what happen if you use s3fs with sftpgo?

**Comment by drakkan** [comment]
I have found a possible improvement, sftpgo could do a HEAD object instead of a list dir in some case

**Comment by fdefilippo** [comment]
Hi, using sftpgo and s3fs is much fuster, same as webdav server and s3fs.

**Comment by drakkan** [comment]
> Hi, using sftpgo and s3fs is much fuster, same as webdav server and s3fs.

yes, cloud backend could be smartest for webdav, it was mainly designed for SFTP/FTP, I'll try to improve it without breaking SFTP/FTP usage, thank you for reporting

**Comment by fdefilippo** [comment]
To speed things up you could add option that overrides mime-types using the file extension as mime-type: application/XXX xxx?

**Comment by drakkan** [comment]
> To speed things up you could add option that overrides mime-types using the file extension as mime-type: application/XXX xxx?

This is only one of the issues, also Stat need improvements

**Comment by drakkan** [comment]
Hi,

please check if the performance is better using the latest Git commit.

A stat is required anyway for each file inside a dir, see [here](https://github.com/golang/net/blob/master/webdav/file.go#L787), This method is called for `PROFIND` if the depth is infinity it resurse all the subdirectories not only the current one.

Thank you

**Comment by fdefilippo** [comment]
something went wrong:

```
{"level":"info","time":"2020-11-02T20:08:14.617","sender":"webdavd","elapsed_ms":12,"method":"PROPFIND","proto":"HTTP/2.0","remote_addr":"127.0.0.1:1470","request_id":"bug5i7jtkmgn87qjpvlg","uri":"https://localhost:2080/test/","user_agent":"curl/7.61.1","error":"permission denied"}
```

```
<?xml version="1.0" encoding="UTF-8"?>
<D:multistatus xmlns:D="DAV:"><D:response><D:href>/test/</D:href><D:propstat><D:prop><D:displayname></D:displayname><D:supportedlock><D:lockentry xmlns:D="DAV:"><D:lockscope><D:exclusive/></D:lockscope><D:locktype><D:write/></D:locktype></D:lockentry></D:supportedlock><D:resourcetype><D:collection xmlns:D="DAV:"/></D:resourcetype><D:getlastmodified>Mon, 02 Nov 2020 19:08:14 GMT</D:getlastmodified></D:prop><D:status>HTTP/1.1 200 OK</D:status></D:propstat></D:response></D:multistatus>Internal Server Error
```

**Comment by drakkan** [comment]
this seems a generic permission denied error, can the same dir be listed using SFTP/FTP? thanks

**Comment by fdefilippo** [comment]
yes it is:
```
$ sftp -o Port=2022 test@localhost
pagamenti@localhost's password:
Connected to test@localhost.
sftp> cd ack
Couldn't stat remote file: Permission denied
sftp> dir
ack     in      out     reject
sftp>
```

no problem with mc ls local/test/ack/

```
$ mc ls caronte/test/ack/|wc -l
943
```

**Comment by drakkan** [comment]
Could it be a virtual permission defined within SFTPGo itself? If the error is returned from the S3 backend you should have a log with the details inside sftpgo logs

**Comment by fdefilippo** [comment]
No virtual permission defined on there 

![chrome_1V0OxbsaKN](https://user-images.githubusercontent.com/16714986/97949844-2ac7c580-1d95-11eb-86f4-1a1cdeb7372b.png)

**Comment by drakkan** [comment]
ok, please post the sftpgo logs then, thank you

**Comment by fdefilippo** [comment]
curl -ksS -X PROPFIND -u test:xxx https://localhost:2080/test/

```
{"level":"debug","time":"2020-11-03T05:51:03.851","sender":"DAV","connection_id":"DAV_buge3drtkmggd80n5n8g","message":"connection added, num open connections: 1"}
{"level":"debug","time":"2020-11-03T05:51:04.033","sender":"dataprovider_sqlite","message":"last login updated for user \"test\""}
{"level":"debug","time":"2020-11-03T05:51:04.099","sender":"DAV","connection_id":"DAV_buge3drtkmggd80n5n8g","message":"transfer added, id: 1, active transfers: 1"}
{"level":"debug","time":"2020-11-03T05:51:04.099","sender":"DAV","connection_id":"DAV_buge3drtkmggd80n5n8g","message":"transfer removed, id: 1 active transfers: 0"}
{"level":"debug","time":"2020-11-03T05:51:04.101","sender":"DAV","connection_id":"DAV_buge3drtkmggd80n5n8g","message":"transfer added, id: 2, active transfers: 1"}
{"level":"debug","time":"2020-11-03T05:51:04.101","sender":"DAV","connection_id":"DAV_buge3drtkmggd80n5n8g","message":"transfer removed, id: 2 active transfers: 0"}
{"level":"debug","time":"2020-11-03T05:51:04.103","sender":"DAV","connection_id":"DAV_buge3drtkmggd80n5n8g","message":"transfer added, id: 3, active transfers: 1"}
{"level":"debug","time":"2020-11-03T05:51:04.108","sender":"DAV","connection_id":"DAV_buge3drtkmggd80n5n8g","message":"transfer removed, id: 3 active transfers: 0"}
{"level":"debug","time":"2020-11-03T05:51:04.110","sender":"DAV","connection_id":"DAV_buge3drtkmggd80n5n8g","message":"error running stat on path \"/ack\": Forbidden: Forbidden\n\tstatus code: 403, request id: 1643E64229564FEF, host id: "}
{"level":"info","time":"2020-11-03T05:51:04.110","sender":"webdavd","elapsed_ms":259,"method":"PROPFIND","proto":"HTTP/2.0","remote_addr":"127.0.0.1:42088","request_id":"buge3drtkmggd80n5n8g","uri":"https://localhost:2080/test/","user_agent":"curl/7.61.1","error":"permission denied"}
{"level":"debug","time":"2020-11-03T05:51:04.110","sender":"DAV","connection_id":"DAV_buge3drtkmggd80n5n8g","message":"connection removed, num open connections: 0"}
```

**Comment by fdefilippo** [comment]
Hi, sorry it is a s3 policy problem, sftp try to use s3.HeadObject but the user lack permission, so I added it. Now is very fast!  

```
time curl -ksS -X PROPFIND -u test:xxx https://localhost:2080/test/ -o /dev/null

real    0m16.790s
```

**Comment by drakkan** [comment]
> Hi, sorry it is a s3 policy problem, sftp try to use s3.HeadObject but the user lack permission, so I added it. Now is very fast!
> 
> ```
> time curl -ksS -X PROPFIND -u test:xxx https://localhost:2080/test/ -o /dev/null
> 
> real    0m16.790s
> ```

It will be even faster, I'll improve `golang/net/webdav` too, thank you

**Comment by fdefilippo** [comment]
Hi, using the windows explorer and trying to copy a file of 18 bytes the msg appears: file too big for the target file system:

![explorer_Tonde6Amp7](https://user-images.githubusercontent.com/16714986/98118121-05c57680-1eab-11eb-999c-83144c63c7c7.png)

```
{"level":"info","time":"2020-11-04T14:34:22.148","sender":"webdavd","elapsed_ms":25,"method":"PROPFIND","proto":"HTTP/2.0","remote_addr":"127.0.0.1:60282","request_id":"buharnjtkmgptsk723mg","uri":"https://localhost:2080/test/out","user_agent":"Microsoft-WebDAV-MiniRedir/10.0.19041"}
{"level":"info","time":"2020-11-04T14:34:22.471","sender":"webdavd","elapsed_ms":262,"method":"PUT","proto":"HTTP/2.0","remote_addr":"127.0.0.1:60282","request_id":"buharnjtkmgptsk723n0","uri":"https://localhost:2080/test/out/datafile.txt","user_agent":"Microsoft-WebDAV-MiniRedir/10.0.19041"}
{"level":"info","time":"2020-11-04T14:34:22.517","sender":"webdavd","elapsed_ms":3,"method":"LOCK","proto":"HTTP/2.0","remote_addr":"127.0.0.1:60282","request_id":"buharnjtkmgptsk723ng","uri":"https://localhost:2080/test/out/datafile.txt","user_agent":"Microsoft-WebDAV-MiniRedir/10.0.19041"}
{"level":"info","time":"2020-11-04T14:34:22.561","sender":"webdavd","elapsed_ms":0,"method":"PROPPATCH","proto":"HTTP/2.0","remote_addr":"127.0.0.1:60282","request_id":"buharnjtkmgptsk723o0","uri":"https://localhost:2080/test/out/datafile.txt","user_agent":"Microsoft-WebDAV-MiniRedir/10.0.19041","error":"webdav: locked"}
{"level":"info","time":"2020-11-04T14:34:22.640","sender":"webdavd","elapsed_ms":4,"method":"HEAD","proto":"HTTP/2.0","remote_addr":"127.0.0.1:60282","request_id":"buharnjtkmgptsk723og","uri":"https://localhost:2080/test/out/datafile.txt","user_agent":"Microsoft-WebDAV-MiniRedir/10.0.19041"}
{"level":"info","time":"2020-11-04T14:34:22.678","sender":"webdavd","elapsed_ms":0,"method":"PUT","proto":"HTTP/2.0","remote_addr":"127.0.0.1:60282","request_id":"buharnjtkmgptsk723p0","uri":"https://localhost:2080/test/out/datafile.txt","user_agent":"Microsoft-WebDAV-MiniRedir/10.0.19041","error":"webdav: locked"}
{"level":"info","time":"2020-11-04T14:34:22.719","sender":"webdavd","elapsed_ms":0,"method":"DELETE","proto":"HTTP/2.0","remote_addr":"127.0.0.1:60282","request_id":"buharnjtkmgptsk723pg","uri":"https://localhost:2080/test/out/datafile.txt","user_agent":"Microsoft-WebDAV-MiniRedir/10.0.19041","error":"webdav: locked"}
{"level":"info","time":"2020-11-04T14:34:23.043","sender":"webdavd","elapsed_ms":279,"method":"PUT","proto":"HTTP/2.0","remote_addr":"127.0.0.1:60282","request_id":"buharnjtkmgptsk723q0","uri":"https://localhost:2080/test/out/datafile.txt","user_agent":"Microsoft-WebDAV-MiniRedir/10.0.19041"}
{"level":"info","time":"2020-11-04T14:34:23.087","sender":"webdavd","elapsed_ms":3,"method":"LOCK","proto":"HTTP/2.0","remote_addr":"127.0.0.1:60282","request_id":"buharnrtkmgptsk723qg","uri":"https://localhost:2080/test/out/datafile.txt","user_agent":"Microsoft-WebDAV-MiniRedir/10.0.19041"}
{"level":"info","time":"2020-11-04T14:34:23.131","sender":"webdavd","elapsed_ms":0,"method":"PROPPATCH","proto":"HTTP/2.0","remote_addr":"127.0.0.1:60282","request_id":"buharnrtkmgptsk723r0","uri":"https://localhost:2080/test/out/datafile.txt","user_agent":"Microsoft-WebDAV-MiniRedir/10.0.19041","error":"webdav: locked"}
{"level":"info","time":"2020-11-04T14:34:23.210","sender":"webdavd","elapsed_ms":3,"method":"HEAD","proto":"HTTP/2.0","remote_addr":"127.0.0.1:60282","request_id":"buharnrtkmgptsk723rg","uri":"https://localhost:2080/test/out/datafile.txt","user_agent":"Microsoft-WebDAV-MiniRedir/10.0.19041"}
{"level":"info","time":"2020-11-04T14:34:23.250","sender":"webdavd","elapsed_ms":0,"method":"PUT","proto":"HTTP/2.0","remote_addr":"127.0.0.1:60282","request_id":"buharnrtkmgptsk723s0","uri":"https://localhost:2080/test/out/datafile.txt","user_agent":"Microsoft-WebDAV-MiniRedir/10.0.19041","error":"webdav: locked"}
{"level":"info","time":"2020-11-04T14:34:23.291","sender":"webdavd","elapsed_ms":0,"method":"DELETE","proto":"HTTP/2.0","remote_addr":"127.0.0.1:60282","request_id":"buharnrtkmgptsk723sg","uri":"https://localhost:2080/test/out/datafile.txt","user_agent":"Microsoft-WebDAV-MiniRedir/10.0.19041","error":"webdav: locked"}
{"level":"info","time":"2020-11-04T14:34:54.081","sender":"webdavd","elapsed_ms":22,"method":"PROPFIND","proto":"HTTP/2.0","remote_addr":"127.0.0.1:60282","request_id":"buharvjtkmgptsk723t0","uri":"https://localhost:2080/test/out","user_agent":"Microsoft-WebDAV-MiniRedir/10.0.19041"}
{"level":"info","time":"2020-11-04T14:34:54.147","sender":"webdavd","elapsed_ms":24,"method":"PROPFIND","proto":"HTTP/2.0","remote_addr":"127.0.0.1:60282","request_id":"buharvjtkmgptsk723tg","uri":"https://localhost:2080/test/out","user_agent":"Microsoft-WebDAV-MiniRedir/10.0.19041"}
{"level":"debug","time":"2020-11-04T14:34:54.187","sender":"connection_failed","client_ip":"127.0.0.1","username":"test","login_type":"password","protocol":"DAV","error":"Forbidden"}
{"level":"debug","time":"2020-11-04T14:34:54.279","sender":"connection_failed","client_ip":"127.0.0.1","username":"test","login_type":"password","protocol":"DAV","error":"Forbidden"}
{"level":"debug","time":"2020-11-04T14:34:54.368","sender":"connection_failed","client_ip":"127.0.0.1","username":"test","login_type":"password","protocol":"DAV","error":"Forbidden"}
{"level":"debug","time":"2020-11-04T14:34:54.455","sender":"connection_failed","client_ip":"127.0.0.1","username":"test","login_type":"password","protocol":"DAV","error":"Forbidden"}
{"level":"info","time":"2020-11-04T14:34:54.499","sender":"webdavd","elapsed_ms":5,"method":"PROPFIND","proto":"HTTP/2.0","remote_addr":"127.0.0.1:60282","request_id":"buharvjtkmgptsk72400","uri":"https://localhost:2080/test","user_agent":"Microsoft-WebDAV-MiniRedir/10.0.19041"}
```

**Comment by drakkan** [comment]
Hi,

please give a try to the [latest commit](https://github.com/drakkan/sftpgo/commit/0a14297b487a9c35095a6659be0608fe4700c815), this issue should be fixed and performance should be better too.

Regarding the directory deletion with storage backends I think cannot be solved

**Comment by fdefilippo** [comment]
> Hi,
> 
> please give a try to the [latest commit](https://github.com/drakkan/sftpgo/commit/0a14297b487a9c35095a6659be0608fe4700c815), this issue with windows exploer should be fixed and performance should be better too. Please let me know if it works better and faster now, thank you

Great! it works fine, we can close this issue. 

Thanks again.

**Comment by drakkan** [comment]
Thank you for confirming,

can you please share the new time for listing you directory containing 943 entry? So I can share this info when I send the webdav performance patch upstream, thanks

**Comment by fdefilippo** [comment]
Hi, here they are 

```
curl -ksS -X PROPFIND -u test:xxZUes3 https://localhost:2080/test/ |xmllint --format -|grep D:href|wc -l
963
```

```
$ time curl -ksS -X PROPFIND -u test:xxZUes3 https://localhost:2080/test/ -o /dev/null   
real    0m0.488s
user    0m0.037s
sys     0m0.013s
```

👍

**Comment by EdenSpire** [comment]
Hello,

This problem still exist on version 2.6.

What should I do?
